django/db/models/fields/related_descriptors.py

```diff
@@ -67,9 +67,17 @@ class Child(Model):
 from django.db import connections, router, transaction
 from django.db.models import Q, signals
 from django.db.models.query import QuerySet
+from django.db.models.query_utils import DeferredAttribute
 from django.utils.functional import cached_property
 
 
+class ForeignObjectDeferredAttribute(DeferredAttribute):
+    def __set__(self, instance, value):
```

Comment:
My only concern about this descriptor it that defining `__set__` makes it a data descriptor which means all getter accesses will go through `DeferredAttribute.__get__` and incur a function call while `DeferredAttribute` is a non-data descriptor and only gets `__get__` called on missing `__dict__` values.

I guess that's a small penality to take in the name of correctness but I thought I'd mention it even if I don't have a better solution for the problem this patch solves.